### PR TITLE
Template injection vulnerability detection in handlebars and pug

### PR DESCRIPTION
**Pull request by IlyasShabi**
### What does this PR do?
Add support to server Template injection using handlebars and pug

### Plugin Checklist

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

File: packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.handlebars.plugin.spec.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                    {{#with (string.sub.apply 0 codelist)}}
+                      {{this}}
+                    {{/with}}
+                  {{/each}}
+                {{/with}}
+              {{/with}}
+            {{/with}}
+          {{/with}}
+          `
+      goodSource = '<p>{{name}}</p>'
```
Comment (uurien) [inline packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.handlebars.plugin.spec.js:34]: For vulnerability detection, the difference about the string is not its content, it is the source. You can use this `<p>{{name}</p>` for vulnerable and not vulnerable tests. It should be tainted in the vulnerable tests and not tainted in not vulnerable tests.

File: packages/datadog-instrumentations/src/handlebars.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+'use strict'
+
+const shimmer = require('../../datadog-shimmer')
+const { channel, addHook } = require('./helpers/instrument')
+
+const handlebarsReadCh = channel('datadog:handlebars:compile:start')
+
+function wrapCompile (compile) {
+  return function () {
```
Comment (simon-id) [inline packages/datadog-instrumentations/src/handlebars.js:9]: fyi in a lot of places in the instrumentation the wrapped function is named, just for clarity like so:
```suggestion
  return function wrappedCompile () {
```

File: packages/datadog-instrumentations/src/handlebars.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    if (handlebarsReadCh.hasSubscribers) {
+      const source = arguments[0]
```
Comment (simon-id) [inline packages/datadog-instrumentations/src/handlebars.js:11]: i usually just use the named function argument, not sure why you're using the arguments var here ?

File: packages/datadog-instrumentations/src/pug.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    }
+    return compile.apply(this, arguments)
+  }
+}
+
+addHook({ name: 'pug', versions: ['>=2.0.4'] }, compiler => {
+  shimmer.wrap(compiler, 'compile', wrapCompile)
+  shimmer.wrap(compiler, 'compileClient', wrapCompile)
+  shimmer.wrap(compiler, 'compileClientWithDependenciesTracked', wrapCompile)
+  return compiler
```
Comment (uurien) [inline packages/datadog-instrumentations/src/pug.js:21]:
```suggestion
  shimmer.wrap(compiler, 'compileClientWithDependenciesTracked', wrapCompile)

  return compiler
```

File: packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.pug.plugin.spec.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        })
+    })
+
+    describe('compileClient', () => {
+      prepareTestServerForIast('template injection analyzer',
+        (testThatRequestHasVulnerability, testThatRequestHasNoVulnerability) => {
+          testThatRequestHasVulnerability(() => {
+            const store = storage.getStore()
+            const iastContext = iastContextFunctions.getIastContext(store)
+            const command = newTaintedString(iastContext, source, 'param', 'Request')
```
Comment (uurien) [inline packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.pug.plugin.spec.js:40]: `command`? Looks like copy&pasted var name, isn't it really `template`?

File: packages/dd-trace/src/appsec/iast/vulnerabilities-formatter/evidence-redaction/sensitive-analyzers/injection-sensitive-analyzer.js
[the remark this answers is not in the document]
Comment (IlyasShabi) [reply packages/dd-trace/src/appsec/iast/vulnerabilities-formatter/evidence-redaction/sensitive-analyzers/injection-sensitive-analyzer.js:1]: I did change the name to `long-content-sensitive-analyzer` but Im open to any other suggestion

File: packages/dd-trace/src/appsec/iast/vulnerabilities-formatter/evidence-redaction/sensitive-handler.js [hunk 6]
```diff
@@ -5,13 +5,13 @@ const vulnerabilities = require('../../vulnerabilities')
...
 const { contains, intersects, remove } = require('./range-utils')
 
-const codeInjectionSensitiveAnalyzer = require('./sensitive-analyzers/code-injection-sensitive-analyzer')
 const commandSensitiveAnalyzer = require('./sensitive-analyzers/command-sensitive-analyzer')
 const hardcodedPasswordAnalyzer = require('./sensitive-analyzers/hardcoded-password-analyzer')
 const headerSensitiveAnalyzer = require('./sensitive-analyzers/header-sensitive-analyzer')
 const jsonSensitiveAnalyzer = require('./sensitive-analyzers/json-sensitive-analyzer')
 const ldapSensitiveAnalyzer = require('./sensitive-analyzers/ldap-sensitive-analyzer')
 const sqlSensitiveAnalyzer = require('./sensitive-analyzers/sql-sensitive-analyzer')
+const injectionSensitiveAnalyzer = require('./sensitive-analyzers/tainted-range-based-sensitive-analyzer')
```
Comment (uurien) [inline packages/dd-trace/src/appsec/iast/vulnerabilities-formatter/evidence-redaction/sensitive-handler.js:14]: As you have renamed the file, you should rename also the var name

File: .github/workflows/appsec.yml [hunk 7]
```diff
@@ -122,7 +122,7 @@ jobs:
   express:
     runs-on: ubuntu-latest
     env:
-      PLUGINS: express|body-parser|cookie-parser|multer
+      PLUGINS: express|body-parser|cookie-parser|multer|handlebars|pug
```
Comment (CarlesDD) [inline .github/workflows/appsec.yml:125]: I think we should not place `handlebars` and `pug` plugin tests in the `express` job.

These tests are not related w/ express, since they are not testing the express integration (they don't do `app.set('view engine', 'pug')` or `app.set('view engine', 'hbs')`). A separate job for these plugin test would be more suitable.

WDYT?

**Comment by CarlesDD** [comment]
PR title should be changed to mention both `handlebars` and `pug`.

File: packages/datadog-instrumentations/src/pug.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+'use strict'
+
+const shimmer = require('../../datadog-shimmer')
+const { channel, addHook } = require('./helpers/instrument')
+
+const pugReadCh = channel('datadog:pug:compile:start')
```
Comment (CarlesDD) [inline packages/datadog-instrumentations/src/pug.js:6]: **Nit**: Var name could be improved by correctly indicating the operation to be published in the DC. In this case, as it will publish a message every time pug compiles, it could be named `pugCompileCh`.

File: packages/datadog-instrumentations/src/handlebars.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
+'use strict'
+
+const shimmer = require('../../datadog-shimmer')
+const { channel, addHook } = require('./helpers/instrument')
+
+const handlebarsReadCh = channel('datadog:handlebars:compile:start')
```
Comment (CarlesDD) [inline packages/datadog-instrumentations/src/handlebars.js:6]: **Nit**: Var name could be improved by correctly indicating the operation to be published in the DC. In this case, as it will publish a message every time handlebars compiles, it could be named `handlebarsCompileCh`.

File: packages/datadog-instrumentations/src/pug.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+      pugReadCh.publish({ source })
+    }
+
+    return compile.apply(this, arguments)
+  }
+}
+
+addHook({ name: 'pug', versions: ['>=2.0.4'] }, compiler => {
+  shimmer.wrap(compiler, 'compile', wrapCompile)
+  shimmer.wrap(compiler, 'compileClient', wrapCompile)
```
Comment (CarlesDD) [inline packages/datadog-instrumentations/src/pug.js:20]: Since `compileClient` [calls internally](https://github.com/pugjs/pug/blob/master/packages/pug/lib/index.js#L363) `compileClientWithDependenciesTracked`, and the latter is also instrumented, are we analyzing twice for SSTi when `compileClient` is called?

File: packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.handlebars.plugin.spec.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+          testThatRequestHasVulnerability(() => {
+            const store = storage.getStore()
+            const iastContext = iastContextFunctions.getIastContext(store)
+            const template = newTaintedString(iastContext, source, 'param', 'Request')
+            lib.compile(template)
+          }, 'TEMPLATE_INJECTION')
+
+          testThatRequestHasNoVulnerability(() => {
+            const template = lib.compile(source)
+            template()
```
Comment (CarlesDD) [inline packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.handlebars.plugin.spec.js:28]: `template()` is invoked in this test (no vuln), but is not invoked in the previous (when checking there is a vuln). Is there any reason for that?

File: packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.pug.plugin.spec.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+          testThatRequestHasVulnerability(() => {
+            const store = storage.getStore()
+            const iastContext = iastContextFunctions.getIastContext(store)
+            const template = newTaintedString(iastContext, source, 'param', 'Request')
+            lib.compile(template)
+          }, 'TEMPLATE_INJECTION')
+
+          testThatRequestHasNoVulnerability(() => {
+            const template = lib.compile(source)
+            template()
```
Comment (CarlesDD) [inline packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.pug.plugin.spec.js:28]: `template()` is invoked in this test (no vuln), but is not invoked in the previous (when checking there is a vuln). Is there any reason for that?

Comment (IlyasShabi) [reply packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.pug.plugin.spec.js:28]: No reason for this just forgot to remove it

File: packages/datadog-instrumentations/src/pug.js (hunk 10 above)
Comment (IlyasShabi) [reply packages/datadog-instrumentations/src/pug.js]: Done

File: packages/datadog-instrumentations/src/handlebars.js (hunk 9 above)
Comment (IlyasShabi) [reply packages/datadog-instrumentations/src/handlebars.js]: Done

File: packages/datadog-instrumentations/src/pug.js (hunk 8 above)
Comment (IlyasShabi) [reply packages/datadog-instrumentations/src/pug.js]: Done

File: .github/workflows/appsec.yml (hunk 7 above)
Comment (IlyasShabi) [reply .github/workflows/appsec.yml]: I previously created separate jobs for `handlebars` and `pug`, but I encountered an error related to express, the only solution I found was to run them with express.
I tried something like `PLUGINS: express|handlebars|pug`  but that ends up running all the express tests again.

I've just pushed my previous commit, and you can see that the `AppSec / template` job is failing.

Comment (CarlesDD) [reply .github/workflows/appsec.yml]: I think the error is not related with express.

In the plugin test, the lib to be tested (hbs, pug) are [required](https://github.com/DataDog/dd-trace-js/pull/4827/files#diff-45d723a65b1a18f2cc2513af6b2acfebad56c4aa9c626be1cf38d22cb4f4925aR12) in the before fn, before calling the method `prepareTestServerForIast`. This method initializes everything necessary to be able to test the plugin, among other things, it initializes the tracer itself, which registers the hook of the require of the library, where we instrument these.

What is happening is that a require of the library is being done before the initialization of the tracer, and therefore, the first test fails because that library is not instrumented (you can see that the second test for the same library but a different version does pass).

The error would be solved by making the library require inside a `beforeEach` inside the `prepareTestServerForIast` cb.

Comment (CarlesDD) [reply .github/workflows/appsec.yml]: Like in this case: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.sequelize.plugin.spec.js#L27

Comment (IlyasShabi) [reply .github/workflows/appsec.yml]: Thanks for this clear explanation, Im gonna fix it

**Comment by CarlesDD** [comment]
In Handlebars there is one more vulnerable point that this PR is not covering: `registerPartial`.

Let's take the following code as an example:

```javascript
  Handlebars.registerPartial('vulnerablePartial', req.query.partial).
  const templateSource = `{{> vulnerablePartial }}`
  const template = Handlebars.compile(templateSource)
```

This code is vulnerable to `SSTi`, but the tests I have run do not detect the vulnerability (and it is logical because `registerPartial` never calls `compile` or `precompile`).

File: packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.pug.plugin.spec.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
+'use strict'
+
+const { prepareTestServerForIast } = require('../utils')
+const { storage } = require('../../../../../datadog-core')
+const iastContextFunctions = require('../../../../src/appsec/iast/iast-context')
+const { newTaintedString } = require('../../../../src/appsec/iast/taint-tracking/operations')
+
+describe('template-injection-analyzer with pug', () => {
```
Comment (CarlesDD) [inline packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.pug.plugin.spec.js:8]: In addition to these already tested methods, we also detected vulnerability when using the `render` method (which internally calls `handleTemplateCache` which in turn calls `compile`).

I think it would be worth adding a test with `render` to check that it still works in future versions, since we are not implementing it implicitly.

Comment (IlyasShabi) [reply packages/dd-trace/test/appsec/iast/analyzers/template-injection-analyzer.pug.plugin.spec.js:8]: Done

File: packages/datadog-instrumentations/src/handlebars.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    if (handlebarsCompileCh.hasSubscribers) {
+      handlebarsCompileCh.publish({ source })
+    }
+
+    return compile.apply(this, arguments)
+  }
+}
+
+function wrapRegisterPartial (handlebarsEnvironment) {
+  return function wrappedCompile (name, partial) {
```
Comment (CarlesDD) [inline packages/datadog-instrumentations/src/handlebars.js:20]:
```suggestion
  return function wrappedRegisterPartial (name, partial) {
```

**Review by CarlesDD — APPROVED** [review]

**Review by uurien — APPROVED** [review]

File: packages/datadog-instrumentations/src/handlebars.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+  return function wrappedCompile (source) {
+    if (handlebarsCompileCh.hasSubscribers) {
+      handlebarsCompileCh.publish({ source })
+    }
+
+    return compile.apply(this, arguments)
+  }
+}
+
+function wrapRegisterPartial (handlebarsEnvironment) {
```
[the remark this answers is not in the document]
Comment (IlyasShabi) [reply packages/datadog-instrumentations/src/handlebars.js]: Done

**Review by CarlesDD — APPROVED** [review]
